### ISSUE TITLE
sql 1.0: unionNode.Columns() crashes when used in subquery RHS of an IN expression

### ISSUE THREAD

**Issue by girishkolantra** [cockroachdb/cockroach#17225]
I created a table using
```
create table node (
   id bigint primary key,
   l_node bigint,
   r_node` bigint
);
```
Then I inserted some data using

```
insert into node (id) values (1);
insert into node (id, l_node) values (2, 1);
insert into node (id, l_node) values (3, 2);
insert into node (id, l_node) values (4, 3);
insert into node (id, l_node) values (5, 4);
insert into node (id, l_node) values (6, 5);
insert into node (id, l_node) values (7, 6);
insert into node (id, l_node) values (8, 7);
insert into node (id, l_node) values (9, 8);
```

I tried to find the leaf node in this tree using the following query
```
select 
	id 
from 
    node 
where 
    id not in (
         select 
              distinct l_node 
         from 
            node 
          where 
            l_node is not null
         union select 
            distinct r_node 
         from 
             node 
          where 
            r_node is not null
    )
```

This query crashed my cluster with three nodes. The following was the stacktrace
```
roach-one_1 | * ERROR: [n1] a panic has occurred!
roach-one_1 | *
roach-one_1 | *
roach-one_1 | * ERROR: [n1] Reported as error 942b1d523f8f47ba82d128ea237fc301
roach-one_1 | *
roach-one_1 | select n1.id from node n1 where id not in (select distinct l_node from node where l_node is not null union select distinct r_node from node where r_node is not null);: runtime error: invalid memory address or nil pointer dereference
roach-one_1 |
roach-one_1 | goroutine 400 [running]:
roach-one_1 | runtime/debug.Stack(0x7f91bf3db8e8, 0xc42034be30, 0x18aad40)
roach-one_1 | /usr/local/go/src/runtime/debug/stack.go:24 +0x79
roach-one_1 | github.com/cockroachdb/cockroach/pkg/util/log.ReportPanic(0x7f91bf3db8e8, 0xc42034be30, 0x1901140, 0xc4201a9900, 0x1)
roach-one_1 | /go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:115 +0x32b
roach-one_1 | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4204eaa00, 0x7f91bf3db8e8, 0xc42034be30)
roach-one_1 | /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:199 +0x99
roach-one_1 | panic(0x1901140, 0xc4201a9900)
roach-one_1 | /usr/local/go/src/runtime/panic.go:489 +0x2cf
roach-one_1 | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve.func1(0xc420e86000, 0x7f91bf3db8e8, 0xc420375dd0)
roach-one_1 | /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:353 +0x74
roach-one_1 | panic(0x1901140, 0xc4201a9900)
roach-one_1 | /usr/local/go/src/runtime/panic.go:489 +0x2cf
roach-one_1 | github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements.func1(0xc4204dd119, 0xa6)
roach-one_1 | /go/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:499 +0x131
roach-one_1 | panic(0x192d620, 0x2749ea0)
roach-one_1 | /usr/local/go/src/runtime/panic.go:489 +0x2cf
roach-one_1 | github.com/cockroachdb/cockroach/pkg/sql.(*unionNode).Columns(0xc420dbc2d0, 0x0, 0xc4201a9810, 0x1)
roach-one_1 | /go/src/github.com/cockroachdb/cockroach/pkg/sql/union.go:140 +0x26
roach-one_1 | github.com/cockroachdb/cockroach/pkg/sql.(*subquery).subqueryTupleOrdering(0xc4212792c0, 0x7f91bf3db8e8, 0xc42157d3e0)
roach-one_1 | /go/src/github.com/cockroachdb/cockroach/pkg/sql/subquery.go:212 +0x49
roach-one_1 | github.com/cockroachdb/cockroach/pkg/sql.(*subquery).doEval(0xc4212792c0, 0x7f91bf3db8e8, 0xc42157d3e0, 0x0, 0x0, 0x0, 0x0)
roach-one_1 | /go/src/github.com/cockroachdb/cockroach/pkg/sql/subquery.go:157 +0x26d
roach-one_1 | github.com/cockroachdb/cockroach/pkg/sql.(*subqueryPlanVisitor).subqueryNode(0xc4205950f0, 0x7f91bf3db8e8, 0xc42157d3e0, 0xc4212792c0, 0x10, 0xc4201a9170)
roach-one_1 | /go/src/github.com/cockroachdb/cockroach/pkg/sql/subquery.go:251 +0xc9
roach-one_1 | github.com/cockroachdb/cockroach/pkg/sql.(*subqueryPlanVisitor).(github.com/cockroachdb/cockroach/pkg/sql.subqueryNode)-fm(0x7f91bf3db8e8, 0xc42157d3e0, 0xc4212792c0, 0xc4212890c8, 0x0)
roach-one_1 | /go/src/github.com/cockroachdb/cockroach/pkg/sql/subquery.go:274 +0x48
roach-one_1 | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).VisitPre(0xc4205dbb00, 0x2788620, 0xc4212792c0, 0xc4212794a0, 0x16, 0xc4212794a0)
roach-one_1 | /go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:448 +0x1c2
roach-one_1 | github.com/cockroachdb/cockroach/pkg/sql/parser.WalkExpr(0x2779520, 0xc4205dbb00, 0x2788620, 0xc4212792c0, 0x2789120, 0xc4212794a0, 0xc4205dba00)
roach-one_1 | /go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/walk.go:552 +0x4c
roach-one_1 | github.com/cockroachdb/cockroach/pkg/sql/parser.(*ComparisonExpr).Walk(0xc42028d580, 0x2779520, 0xc4205dbb00, 0xc421289201, 0x2788ae0)
roach-one_1 | /go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/walk.go:178 +0xa8
roach-one_1 | github.com/cockroachdb/cockroach/pkg/sql/parser.WalkExpr(0x2779520, 0xc4205dbb00, 0x2788ae0, 0xc42028d580, 0xc421289270, 0x2788ae0, 0xc4205dba01)
roach-one_1 | /go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/walk.go:555 +0x15a
```

I am using cockroachdb build:      CCL v1.0.3 @ 2017/07/06 17:46:06 (go1.8.3)

**Comment by vivekmenezes** [comment]
Thanks for reporting this issue. We'll attempt to fix it in the next release.

**Comment by jordanlewis** [comment]
@girishkolantra, I cannot reproduce this bug on a recent build - it was probably fixed already. You should be able to successfully run your query on a recent alpha, or wait for 1.1.

**Comment by vivekmenezes** [comment]
https://github.com/cockroachdb/cockroach/pull/16241 fixed this issue. The removal of the Columns() function in Union fixed it. Assigning to @knz to close

**Comment by knz** [comment]
There were multiple changes to this code in 1.1 indeed, but no particular PR was responsible for fixing the bug you observe - it's the combination of changes that made the problem go away.

The problem still exists in 1.0.x however and will occur in any query with an expression of  the form `x IN (... UNION ...)`. The question is whether we want to fix it in the 1.0.x, so close to the release of 1.1.

@girishkolantra would it be acceptable for you to migrate your cluster to use v1.1 instead?

**Comment by knz** [comment]
Fixed #17933
